File: zuup-tauri/src/api/download.ts
import { invoke } from '@tauri-apps/api/core';
import type { DownloadInfo, DownloadStats } from '@/types';

export class DownloadAPI {
  static async addDownload(
    url: string, 
    options?: { download_path?: string; filename?: string }
  ): Promise<string> {
    return invoke('add_download', {
      url,
      downloadPath: options?.download_path,
      filename: options?.filename,
    });
  }

  static async pauseDownload(id: string): Promise<void> {
    return invoke('pause_download', { id });
  }

  static async resumeDownload(id: string): Promise<void> {
    return invoke('resume_download', { id });
  }

  static async cancelDownload(id: string): Promise<void> {
    return invoke('cancel_download', { id });
  }

  static async removeDownload(id: string): Promise<void> {
    return invoke('remove_download', { id });
  }

  static async getDownloads(): Promise<DownloadInfo[]> {
    return invoke('get_downloads');
  }

  static async getDownloadStats(): Promise<DownloadStats> {
    return invoke('get_download_stats');
  }
}

File: zuup-tauri/src-tauri/src/commands/download.rs
use tauri::State;
use crate::services::{DownloadService, download::{AddDownloadRequest, DownloadInfo, DownloadStats}};

#[tauri::command]
pub async fn add_download(
    download_service: State<'_, DownloadService>,
    url: String,
    download_path: Option<String>,
    filename: Option<String>,
) -> Result<String, String> {
    let request = AddDownloadRequest {
        url,
        download_path,
        filename,
    };
    
    download_service.add_download(request)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn pause_download(
    download_service: State<'_, DownloadService>,
    id: String,
) -> Result<(), String> {
    download_service.pause_download(&id)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn resume_download(
    download_service: State<'_, DownloadService>,
    id: String,
) -> Result<(), String> {
    download_service.resume_download(&id)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn cancel_download(
    download_service: State<'_, DownloadService>,
    id: String,
) -> Result<(), String> {
    download_service.cancel_download(&id)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn remove_download(
    download_service: State<'_, DownloadService>,
    id: String,
) -> Result<(), String> {
    download_service.remove_download(&id)
        .map_err(|e| e.to_string())
}

#[tauri::command]
pub async fn get_downloads(
    download_service: State<'_, DownloadService>,
) -> Result<Vec<DownloadInfo>, String> {
    Ok(download_service.get_downloads())
}

#[tauri::command]
pub async fn get_download_stats(
    download_service: State<'_, DownloadService>,
) -> Result<DownloadStats, String> {
    Ok(download_service.get_download_stats())
}